tests/testthat/test_distance.R

```diff
@@ -0,0 +1,37 @@
+x <- matrix(c(1:5, 1:5), ncol = 2)
+y <- matrix(c(1:5, 5:1), ncol = 2)
+
+test_that("ndotproduct", {
```

Comment:
is there a way that we could cross-check with external results (i.e. results from another/original implementation of the methods)?

Reply:
That would be really great. But I only aware of our implementation in `MSnbase` and the one in `OrgMassSpecR`. The latter produce (of course) different results for the dotproduct. Especially the weighting isn't available in any tools I know:
https://github.com/OrgMassSpec/OrgMassSpecR/blob/aa0e4ee927b7f7051efa6400ad6f3fae8539f7c8/OrgMassSpecR/R/SpectrumSimilarity.R#L42
```r
similarity_score <- as.vector((u %*% v) / (sqrt(sum(u^2)) * sqrt(sum(v^2))))
```
vs our:
https://github.com/rformassspectrometry/MsCoreUtils/blob/b1e1ad095197d6ea558662348504202d8321a099/R/distance.R#L79-L84

Which (if `m = 0` and `n = 0.5`) boils down to:
```r
sum(sqrt(u) * sqrt(v))^2L / (sum(u) * sum(v))
# vs OrgMassSpecR
sum(u * v) / (sqrt(sum(u^2)) * sqrt(sum(v^2)))
```
As you see already this "simple" normalised dotproduct is hard to implement correctly.

@tobiasko, @tnaake are you aware of any software/web service were we can double check our results?

Reply:
For the spectral contrast angle (SCA) I found this paper:

https://link.springer.com/content/pdf/10.1016/S1044-0305(01)00327-0.pdf

but they do simulations and list means/medians in the paper. But there is some basic geometric stuff in there that might help!

Reply:
Hmmm...what about the following: Define s (actual position and intensity values do not matter), generate s' that is orthogonal in all dimension. Could be done by a function f(s) -> s'. The geometric definition of SCA would require that SCA of s vs. s` is 90 degree.

Reply:
@tobiasko I am not sure that this is working because we the `dotproduct` is part of the `spectralangle` and we use a normalised dotpoduct here. So may we have to implement an unnoramlized dotproduct and spectral angle first?

Reply:
I guessed already that it might be tricky. Suggestion: add a `@note` telling that the methods are implemented as described in xxx but that due to the lack of a reference implementation we can not check that the results are identical?

Reply:
Is there already a function in `Spectra` that does the m/z binning? Using the def. from [Methods to Calculate Spectrum Similarity](https://link.springer.com/protocol/10.1007%2F978-1-4939-6740-7_7) one first derives a binned spectrum from the measured data, before calc. for instance the dot product:

"To create such vectors from acquired MS/MS spectra, each spectrum is divided into the same number of n bins, with n either set to a fixed-value [15, 22], or determined based on fragment ion tolerance [17]. Each bin is assigned a certain weight, calculated by summing up all peak intensities in that bin [22], or set to the highest peak intensity in that bin [17]. These binned spectra can now be considered as two vectors of equal dimensionality n, and can be matched against each other by the (normalized) dot product."

If I get this correctly, the binning makes sure that both spectra have aligned dimensions. How would one represent such a binned spectrum? The vectors will most likely be very sparse, at least at small bin sizes. Ideas?

Reply:
@sgibb @jorainer Since there was no reply here is my version of a bin function. The function forces all m/z values of a spectrum (here a matrix) on a tuneable grid (default 1 u) and aggregates intensities by summation:

```
mzbin <- function(x, digits = 0){
  stopifnot(class(x) == "matrix", c("mz", "int") %in% dimnames(x)[[2]] )
  x[,"mz"] <- round(x[,"mz"], digits = digits)
  aggregate(int ~ mz, x, FUN = sum)
}
```

A basic demo:

```
> S1
        mz int
[1,] 100.0   1
[2,] 101.2   2
[3,] 101.4   3
> mzbin(S1)
   mz int
1 100   1
2 101   5
```
Simple, but pretty effective and it should work for big data sets (collections of Spectra)! What do you think?

Reply:
Based on the above one can calc. a dot product for spectra in sparse matrix notation:
```
## dot product for vectors along m/z index (position)
dotProduct <- function(x, y){
  stopifnot(class(x) == "data.frame", class(y) == "data.frame",
            colnames(x) == c("mz", "int"), colnames(y) == c("mz", "int"))
  z <- merge(x, y, by = 'mz', all = TRUE) #full outer join
  sum(apply(z[,c("int.x", "int.y")], 1, prod), na.rm = TRUE) #sum of position-wise product
}
```
a small test:

```
> ## generate simple vectors along index 1, 2, ..., n
> v1 <- c(1, 2, 3)
> v2 <- c(4, 5, 6)
> v1 %*% v2 #dotp product along index
     [,1]
[1,]   32
> 
> ## create sparse spectrum along mz index in matrix notation
> S1 <- cbind(mz = c(100, 101.2, 102), int = v1)
> S2 <- cbind(mz = c(100, 100.91, 102), int = v2)
> v1 %*% v2 == dotProduct(mzbin(S1), mzbin(S2))
     [,1]
[1,] TRUE
```
:-)

Reply:
The normalized dot product shoudl be:

```
> ## norm. dot productn for simple vectors
> x <- 1:3
> y <- 1:3
> 1 == (x %*% y)/(sqrt(sum(x^2)) * sqrt(sum(y^2)))
     [,1]
[1,] TRUE
```
same as `x%*%y / norm(x) * norm(y)`

Reply:
and here is the norm. dot product along the lines above:

```
## norm. dot product for vectors along m/z dimension
normDotProduct <- function(x, y){
  stopifnot(class(x) == "data.frame", class(y) == "data.frame",
            colnames(x) == c("mz", "int"), colnames(y) == c("mz", "int"))
  z <- merge(x, y, by = 'mz', all = TRUE) #full outer join
  dp <- sum(apply(z[,c("int.x", "int.y")], 1, prod), na.rm = TRUE) #sum of position-wise product
  normx <- sqrt(sum(x[,"int"]^2))
  normy <- sqrt(sum(y[,"int"]^2))
  dp / (normx * normy)
}
```
test for edge cases:

```
> ## test
> ## case 1: all peaks match on 1 u grid 
> S1 <- cbind(mz = c(100, 101.2, 102), int = 1:3)
> S2 <- cbind(mz = c(100, 100.91, 102), int = 1:3)
> 1 == normDotProduct(mzbin(S1), mzbin(S2))
[1] TRUE
> 
> ## case 2 : no peaks match on 1 u grid
> S2 <- cbind(mz = c(200, 200.91, 202), int = 1:3)
> 0 == normDotProduct(mzbin(S1), mzbin(S2))
[1] TRUE
```
This is far more powerful than [R/dotproduct.R](https://github.com/rformassspectrometry/MsCoreUtils/blob/master/R/dotproduct.R)
which just warns if peaks are not aligned and still computes a value.

Reply:
A more general question @sgibb : Why do you want to place these functions outside of the `Spectra` package (MSCoreUtils)? Why not simply in `Spectra`? A "." version for internal usage and versions that take `Spectra` objects as arguments (exported) and are fully vectorised?

Reply:
@tobiasko: thanks for your code suggestions but we already have a [binning](https://rformassspectrometry.github.io/MsCoreUtils/reference/binning.html) and an outer [join](https://rformassspectrometry.github.io/MsCoreUtils/reference/matching.html).

`dotproduct` etc. are needed in `Chromatograms` as well. That's why we put it here. We already have a `compareSpectra` in `Spectra`.

> This is far more powerful than R/dotproduct.R

Why? And `R/dotproduct.R` is obsolete with this PR.

Reply:
Why? Because...
```
> ## copied from
> ## https://github.com/rformassspectrometry/MsCoreUtils/blob/distance/R/distance.R
> 
> .weightxy <- function(x, y, m = 0, n = 0.5) {
+   x ^ m * y ^ n
+ }
> 
> ndotproduct <- function(x, y, m = 0L, n = 0.5, na.rm = TRUE) {
+   wx <- .weightxy(x[, 1L], x[, 2L], m, n)
+   wy <- .weightxy(y[, 1L], y[, 2L], m, n)
+   sum(wx * wy, na.rm = na.rm)^2L /
+     (sum(wx^2L, na.rm = na.rm) * sum(wy^2L, na.rm = na.rm))
+ }
> 
> ## test on case with no shared peaks
> x <- matrix(c(6:10, 1:5), ncol = 2, dimnames = list(c(), c("mz", "intensity")))
> y <- matrix(c(1:5, 1:5), ncol = 2, dimnames = list(c(), c("mz", "intensity")))
> ndotproduct(x, y)
[1] 1
> # :-(
```
and that should be zero by definition!? Please compare to case 2 above.

Reply:
These is expected. As you mentioned above all distance/similarity measurements require binned spectra. The input in your example is not aligned but `ndotproduct` assumes it is (and for that an intensity of `1:5` against an intensity `1:5` should return a similarity of `1`). So a fair comparison would be:
```r
joinndotproduct <- function(x, y) {
    j <- join(x[, "mz"], y[, "mz"], type = "outer")
    ndotproduct(x[j$x,], y[j$y,])
}

joinndotproduct(x, y)
# [1] 0
```
That is exactly what `Spectra::compareSpectra` is doing. `compareSpectra` is more powerful because it allows the user to select a joining function (outer/inner/left/right/(graph based if the other PR will be accepted)/user provided) and a distance/similarity function (one of these that are part of this PR or any other function the user provides).

We try to build functions for just one task (following the unix philosophy: doing on thing and do it well. Your `normDotProduct` does at least 2 task: joining and distance measurement).

Reply:
As already mentioned above my problem with the dotproduct is the following:

The general definition of a normalized dotproduct is:
```
sum(u * v) / (sqrt(sum(u^2)) * sqrt(sum(v^2))
```
but in the literature (e.g. Stein and Scott 1994) it is 
```
sum(sqrt(u) * sqrt(v))^2L / (sum(u) * sum(v)))
```
(for the default weighting with m = 0, and n = 0.5)

This will definitively yield different results.

Reply:
Also providing some background info: we did separate the binning/peak matching and similarity calculation steps because we wanted to give the users the choice to select a specific method for each of these steps (i.e. some people might prefer binning the data while others might want to match peaks based on difference of their m/z). The similarity calculations are performed on the matched (or *aligned* spectra). It would thus also be possible to use different, base R, similarity calculation functions.

For the reason to have these in `MsCoreUtils` and not in `Spectra`: some/many of these functions will be re-used in the `Chromatograms` package.